tough/src/editor.rs

```diff
@@ -0,0 +1,660 @@
+        let signed_snapshot = self
+            .build_snapshot(&signed_targets)
+            .and_then(|snapshot| self.sign_role(snapshot, &root_keys, &rng))?;
+        let signed_timestamp = self
+            .build_timestamp(&signed_snapshot)
+            .and_then(|timestamp| self.sign_role(timestamp, &root_keys, &rng))?;
+
+        Ok(SignedRepository {
+            root: self.signed_root.clone(),
+            targets: signed_targets,
+            snapshot: signed_snapshot,
+            timestamp: signed_timestamp,
+        })
+    }
+
+    // Gets the corresponding keys from Root (root.json) for the given KeySources
+    fn get_root_keys(&self, keys: Vec<Box<dyn KeySource>>) -> Result<RootKeys> {
+        let root = &self.signed_root.signed.signed;
+        let mut root_keys = RootKeys::new();
+
+        for source in keys {
+            let key_pair = source.as_sign().context(error::KeyPairFromKeySource)?;
+
+            if let Some(key_id) = root.key_id(key_pair.as_ref()) {
+                root_keys.insert(key_id, key_pair);
+            }
+        }
+        ensure!(!root_keys.is_empty(), error::KeysNotFoundInRoot);
+        Ok(root_keys)
+    }
+
+    // Build the Targets struct
+    fn build_targets(&self) -> Result<Targets> {
+        let version = self.targets_version.context(error::Missing {
+            field: "targets version",
+        })?;
+        let expires = self.targets_expires.context(error::Missing {
+            field: "targets expiration",
+        })?;
+
+        // BEWARE!!! We are allowing targets to be empty! While this isn't
+        // the most common use case, it's possible this is what a user wants.
+        // If it's important to have a non-empty targets, the object can be
+        // inspected by the calling code.
+        let mut targets: HashMap<String, Target> = HashMap::new();
+        if let Some(ref existing_targets) = self.existing_targets {
+            targets.extend(existing_targets.clone());
+        }
+        if let Some(ref new_targets) = self.new_targets {
+            targets.extend(new_targets.clone());
+        }
+        //ensure!(!targets.is_empty(), error::NoTargets);
+
+        let _extra = self.targets_extra.clone().unwrap_or_else(HashMap::new);
+        Ok(Targets {
+            spec_version: SPEC_VERSION.to_string(),
+            version,
+            expires,
+            targets,
+            _extra,
+        })
+    }
+
+    // Build the Snapshot struct
+    fn build_snapshot(&self, signed_targets: &SignedRole<Targets>) -> Result<Snapshot> {
+        let version = self.snapshot_version.context(error::Missing {
+            field: "snapshot version",
+        })?;
+        let expires = self.snapshot_expires.context(error::Missing {
+            field: "snapshot expiration",
+        })?;
+        let _extra = self.snapshot_extra.clone().unwrap_or_else(HashMap::new);
+
+        let targets_meta = Self::snapshot_meta(signed_targets);
```

Comment:
Sorry if this is a dumb question... why do we use `snapshot_meta` to make a variable called `targets_meta` associated with `targets.json`, and below we use `timestamp_meta` to make a variable called `snapshot_meta` associated with `snapshot.json`?  Naively they seem to be used backwards... :)

I think we could use a little more documentation on/in these build methods.  It's complicated stuff.

Reply:
Not a dumb question at all. I think the confusion comes from the variable name (suggestions welcome). Snapshot contains the filename and hashes of the signed `Targets` and `Root` data. Further details below.

The `Snapshot` struct has a member `meta`. `Snapshot.meta` contains a `HashMap<String, SnapshotMeta>`.  This `Hashmap` is where the data about the `Targets/Root` is stored. In practice it looks something like:

```
{
   ...,
   "meta":{
      "root.json":{
         "length":1,
         "hashes":{
            "sha256":"1234"
         },
         "version":1
      },
      "targets.json":{
         "length":2,
         "hashes":{
            "sha256":"5678"
         },
         "version":1588449897
      }
   }
}
```

Reply:
Maybe a comment above this like `// Snapshot stores metadata about targets and root.`  And below in the other function, `// Targets stores metadata about snapshot.` or something..?